### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by shumYevh**

### CODE REVIEW

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    private final UserRepository userRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public ShoppingCart createShoppingCart(ShoppingCart shoppingCart) {
+        return shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Transactional
```
Comment (Vasul-Pankiv) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]: why you make get method transactional?

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java (hunk 1 above)
Comment (shumYevh) [reply src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]: LazyInit

Comment (shumYevh) [reply src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]: I fix this by Join fetch in the next commit

File: src/main/java/org/example/bookweb/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping cart management",
+        description = "Endpoints for managing shopping cart with cart items")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/api/cart")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/controller/ShoppingCartController.java]:
```suggestion
@RequestMapping("/cart")
```
Add `server.servlet.context-path=/api` in application.properties

File: src/main/java/org/example/bookweb/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping
+    @Operation(summary = "Get Shopping Cart",
+            description = "Get Shopping Cart with Cart Items")
+    public ShoppingCartResponseDto getShoppingCart() {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartResponseDto getShoppingCart(Authentication authentication) {
```

File: src/main/java/org/example/bookweb/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        String userName = SecurityContextHolder.getContext().getAuthentication().getName();
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/controller/ShoppingCartController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return shoppingCartService.getShoppingCart(user.getId());
```

File: src/main/java/org/example/bookweb/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        return shoppingCartService.getShoppingCart(userName);
+    }
+
+    @PostMapping
+    @Operation(summary = "Add Cart Item to Shopping cart",
+            description = "Add Cart Item with dto to personal Shopping Cart")
+    public ShoppingCartResponseDto addCartItem(@RequestBody @Valid AddCartItemDto addCartItemDto) {
+        String email = SecurityContextHolder.getContext().getAuthentication().getName();
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/controller/ShoppingCartController.java]: the same as above.
Check other places

File: src/main/java/org/example/bookweb/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/cart-items/{cartItemId}")
+    @Operation(summary = "Delete Cart Item",
+            description = "Delete Cart Item from Shopping Cart")
+    public void deleteCartItem(@PathVariable Long cartItemId) {
+        String email = SecurityContextHolder.getContext().getAuthentication().getName();
+        shoppingCartService.removeCarItemFromCart(email, cartItemId);
+    }
+
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/controller/ShoppingCartController.java]:
```suggestion
```

File: src/main/java/org/example/bookweb/dto/shopping/cart/AddCartItemDto.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package org.example.bookweb.dto.shopping.cart;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class AddCartItemDto {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/dto/shopping/cart/AddCartItemDto.java]:
```suggestion
public class CartItemRequestDto {
```
You can use java record for DTO

File: src/main/java/org/example/bookweb/mapper/CartItemMapper.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package org.example.bookweb.mapper;
+
+import org.example.bookweb.config.MapperConfig;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/mapper/CartItemMapper.java:10]:
```suggestion
public interface CartItemMapper {
    CartItem toCartItem(CartItemRequestDto cartItemDto);

    @Mapping(source = "book.id", target = "bookId")
    @Mapping(source = "book.title", target = "bookTitle")
    CartItemResponseDto toDto(CartItem cartItem);
```

File: src/main/java/org/example/bookweb/mapper/ShoppingCartMapper.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+import org.example.bookweb.config.MapperConfig;
+import org.example.bookweb.dto.shopping.cart.CartItemResponseDto;
+import org.example.bookweb.dto.shopping.cart.ShoppingCartResponseDto;
+import org.example.bookweb.models.CartItem;
+import org.example.bookweb.models.ShoppingCart;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/mapper/ShoppingCartMapper.java]:
```suggestion
@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
```

File: src/main/java/org/example/bookweb/mapper/ShoppingCartMapper.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+public interface ShoppingCartMapper {
+
+    @Mapping(source = "user.id", target = "userId")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/mapper/ShoppingCartMapper.java:12]:
```suggestion
    @Mapping(source = "user.id", target = "userId")
    ShoppingCartResponseDto toDto(ShoppingCart shoppingCart);
```

File: src/main/java/org/example/bookweb/mapper/ShoppingCartMapper.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    @Mapping(source = "cartItems", target = "cartItems")
+    ShoppingCartResponseDto toShoppingCartResponseDto(ShoppingCart shoppingCart);
+
+    @Mapping(source = "book.id", target = "bookId")
+    @Mapping(source = "book.title", target = "bookTitle")
+    CartItemResponseDto toCartItemResponseDto(CartItem cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/mapper/ShoppingCartMapper.java]: remove

File: src/main/java/org/example/bookweb/models/CartItem.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+@Getter
+@Setter
+@Entity
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/models/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/org/example/bookweb/models/CartItem.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    @JoinColumn(nullable = false, name = "cart_id")
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/models/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/org/example/bookweb/models/ShoppingCart.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinColumn(nullable = false, name = "user_id")
+    private User user;
+    @OneToMany (mappedBy = "shoppingCart", orphanRemoval = true)
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/models/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/org/example/bookweb/models/ShoppingCart.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+@Getter
+@Setter
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/models/ShoppingCart.java]: You can use
```suggestion
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/org/example/bookweb/repository/shopping/cart/CartItemRepository.java [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.Optional;
+import org.example.bookweb.models.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+
+public interface CartItemRepository extends
+        JpaRepository<CartItem, Long>,
+        JpaSpecificationExecutor<CartItem> {
+    Optional<CartItem> findCartItemById(Long cartId);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/repository/shopping/cart/CartItemRepository.java]: It's better to use 
```suggestion
    Optional<CartItem> findByIdAndShoppingCartId(Long itemId, Long cartId);
```

File: src/main/java/org/example/bookweb/repository/shopping/cart/ShoppingCartRepository.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends
+        JpaRepository<ShoppingCart, Long>,
+        JpaSpecificationExecutor<ShoppingCart> {
+
+    @Query("SELECT cart FROM ShoppingCart cart LEFT JOIN FETCH cart.cartItems")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/repository/shopping/cart/ShoppingCartRepository.java]:
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
```

File: src/main/java/org/example/bookweb/repository/shopping/cart/ShoppingCartRepository.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    ShoppingCart findByUser(User userId);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/repository/shopping/cart/ShoppingCartRepository.java]:
```suggestion
    ShoppingCart findByUser(Long userId);
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public ShoppingCart createShoppingCart(ShoppingCart shoppingCart) {
+        return shoppingCartRepository.save(shoppingCart);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java:32]:
```suggestion
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.setUser(user);
        shoppingCartRepository.save(shoppingCart);
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public ShoppingCart createShoppingCart(ShoppingCart shoppingCart) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCart createShoppingCart(User user) {
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+        return shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart(String email) {
+        return shoppingCartMapper.toShoppingCartResponseDto(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
       return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(userId));
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    public ShoppingCart createShoppingCart(ShoppingCart shoppingCart) {
+        return shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart(String email) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartResponseDto getShoppingCart(Long userId) {
```

File: src/main/java/org/example/bookweb/repository/shopping/cart/ShoppingCartRepository.java [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+
+    @Query("SELECT cart FROM ShoppingCart cart "
+            + "JOIN FETCH cart.cartItems "
+            + "JOIN cart.user u "
+            + "WHERE u.email = :email")
+    ShoppingCart findShoppingCartByUserEmail(@Param("email") String email);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/repository/shopping/cart/ShoppingCartRepository.java]: You can remove this method and find shopping cart by userId

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+        return shoppingCartMapper.toShoppingCartResponseDto(
+                shoppingCartRepository.findShoppingCartByUserEmail(email));
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(AddCartItemDto addCartItemDto, String email) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserEmail(email);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart(String email) {
+        return shoppingCartMapper.toShoppingCartResponseDto(
+                shoppingCartRepository.findShoppingCartByUserEmail(email));
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(AddCartItemDto addCartItemDto, String email) {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartResponseDto addCartItem(AddCartItemDto addCartItemDto, Long userId) {
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        return shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCartRepository.save(shoppingCart));
+    }
+
+    @Override
+    public ShoppingCartResponseDto updateCartItem(String email,
+                                                  Long cartItemId,
+                                                  UpdateCartItemDto updateCartItemDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
                .map(item -> {
                    item.setQuantity(updateCartItemDto.quantity());
                    return item;
                }).orElseThrow(() -> new EntityNotFoundException(
                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
                ));
        cartItemRepository.save(cartItem);
        return shoppingCartMapper.toDto(cart);
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+            cartItem.setQuantity(addCartItemDto.getQuantity());
+            cartItemRepository.save(cartItem);
+            shoppingCart.getCartItems().add(cartItem);
+        }
+        return shoppingCartMapper
+                .toShoppingCartResponseDto(shoppingCartRepository.save(shoppingCart));
+    }
+
+    @Override
+    public ShoppingCartResponseDto updateCartItem(String email,
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartResponseDto updateCartItem(Long userId,
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+        );
+        cartItem.setQuantity(updateCartItemDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toShoppingCartResponseDto(
+                shoppingCartRepository.findShoppingCartByUserEmail(email));
+    }
+
+    @Override
+    public void removeCarItemFromCart(String email, Long cartItemId) {
+        CartItem cartItem = cartItemRepository.findCartItemById(cartItemId).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, shoppingCart.getId())....
```

File: src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+import org.example.bookweb.models.CartItem;
+import org.example.bookweb.models.ShoppingCart;
+import org.example.bookweb.repository.BookRepository;
+import org.example.bookweb.repository.shopping.cart.CartItemRepository;
+import org.example.bookweb.repository.shopping.cart.ShoppingCartRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/shopping/cart/ShoppingCartServiceImpl.java:22]: Add `@Tramsactional` where is needed

File: src/main/java/org/example/bookweb/service/user/UserServiceImpl.java [hunk 30]
```diff
@@ -34,6 +37,9 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         newUser.setPassword(passwordEncoder.encode(requestDto.getPassword()));
         newUser.setRoles(Set.of(roleRepository.getByRole(Role.RoleName.USER)));
         User savedUser = userRepository.save(newUser);
+        ShoppingCart shoppingCart = new ShoppingCart();
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookweb/service/user/UserServiceImpl.java]: Refactor the logic according to the comments in ShoppingCartServiceImpl class

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/org/example/bookweb/models/ShoppingCart.java (hunk 15 above)
Comment (okuzan) [reply src/main/java/org/example/bookweb/models/ShoppingCart.java]: not fixed

File: src/main/java/org/example/bookweb/models/ShoppingCart.java (hunk 14 above)
Comment (okuzan) [reply src/main/java/org/example/bookweb/models/ShoppingCart.java]: not fixed

File: src/main/java/org/example/bookweb/repository/shopping/cart/CartItemRepository.java (hunk 16 above)
Comment (okuzan) [reply src/main/java/org/example/bookweb/repository/shopping/cart/CartItemRepository.java]: not fixed

**Review by okuzan — CHANGES_REQUESTED** [review]
make sure to fix all comments, maybe you skipped those?
image

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**